// react-three-map/dist/es/main17.mjs
import { jsx, Fragment } from "react/jsx-runtime";
import { createRoot, _roots } from "@react-three/fiber";
import { useState, useEffect } from "react";
import { events } from "./main20.mjs";
import { setCoords, useSetRootCoords } from "./main15.mjs";
import { useFunction } from "./main10.mjs";
import { initR3M } from "./main13.mjs";
function useRoot(fromLngLat, map, { frameloop, longitude, latitude, altitude, ...props }) {
  const [{ root, useThree, canvas, r3m }] = useState(() => {
    const canvas2 = map.getCanvas();
    const gl = canvas2.getContext("webgl2") || canvas2.getContext("webgl");
    const root2 = createRoot(canvas2);
    root2.configure({
      dpr: window.devicePixelRatio,
      events,
      ...props,
      frameloop: "never",
      gl: {
        context: gl,
        autoClear: false,
        antialias: true,
        ...props == null ? void 0 : props.gl
      },
      onCreated: (state) => {
        state.gl.forceContextLoss = () => {
        };
      },
      camera: {
        matrixAutoUpdate: false,
        near: 0
      },
      size: {
        width: canvas2.clientWidth,
        height: canvas2.clientHeight,
        top: canvas2.offsetTop,
        left: canvas2.offsetLeft,
        updateStyle: false,
        ...props == null ? void 0 : props.size
      }
    });
    const store = _roots.get(canvas2).store;
    const r3m2 = initR3M({ map, fromLngLat, store });
    setCoords(store, { longitude, latitude, altitude });
    if (frameloop === "demand") {
      store.setState({
        frameloop,
        invalidate: () => {
          map.triggerRepaint();
        }
      });
    }
    return { root: root2, useThree: store, map, canvas: canvas2, r3m: r3m2 };
  });
  const onResize = useFunction(() => {
    const { setDpr, setSize } = useThree.getState();
    setDpr(window.devicePixelRatio);
    setSize(
      canvas.clientWidth,
      canvas.clientHeight,
      false,
      canvas.offsetTop,
      canvas.offsetLeft
    );
  });
  const onRemove = useFunction(() => {
    root.unmount();
  });
  useSetRootCoords(useThree, { longitude, latitude, altitude });
  useEffect(() => {
    if (frameloop !== "demand")
      return;
    const setState = useThree.setState;
    const { invalidate } = useThree.getState();
    setState({
      frameloop,
      invalidate: () => {
        map.triggerRepaint();
      }
    });
    return () => {
      setState({ frameloop: "never", invalidate });
    };
  }, [frameloop]);
  useEffect(() => {
    map.on("resize", onResize);
    return () => {
      map.off("resize", onResize);
    };
  }, []);
  useEffect(() => {
    root.render(/* @__PURE__ */ jsx(Fragment, { children: props.children }));
  }, [props.children]);
  return { onRemove, useThree, r3m };
}
export {
  useRoot
};
